nav2_mppi_controller/src/critics/obstacles_critic.cpp

```diff
@@ -120,26 +119,31 @@ void ObstaclesCritic::score(CriticData & data)
     near_goal = true;
   }
 
-  auto && raw_cost = xt::xtensor<float, 1>::from_shape({data.costs.shape(0)});
-  auto && repulsive_cost = xt::xtensor<float, 1>::from_shape({data.costs.shape(0)});
+  Eigen::ArrayXf raw_cost = Eigen::ArrayXf::Zero(data.costs.size());
+  Eigen::ArrayXf repulsive_cost = Eigen::ArrayXf::Zero(data.costs.size());
 
-  const size_t traj_len = data.trajectories.x.shape(1);
+  const unsigned int traj_len = data.trajectories.x.cols();
+  const unsigned int batch_size = data.trajectories.x.rows();
   bool all_trajectories_collide = true;
-  for (size_t i = 0; i < data.trajectories.x.shape(0); ++i) {
+  const auto & traj = data.trajectories;
+
+  // Default layout in eigen is column-major, hence accessing elements
+  // in column-major fashion to utilize L1 cache as much as possible
+  for(unsigned int i = 0; i != traj_len; i++) {
```

Comment:
I think this should be looked at again, I don't think this is actually a 1:1 reimplementation. The behavior with `raw_cost`, `traj_cost` dont appear to me to be the same

Reply:
Yes, It was not the same exactly. I tried column-major looping to gain some performance by contiguous buffer access. But I've reverted now to the original implementation as behaviour was not same.